src/coreclr/jit/lsraarm64.cpp

```diff
@@ -1984,21 +1985,7 @@ int LinearScan::BuildHWIntrinsic(GenTreeHWIntrinsic* intrinsicTree, int* pDstCou
                 // Nothing needs to be done
             }
 
-            GenTree* ops[] = {intrinEmb.op1, intrinEmb.op2, intrinEmb.op3};
-            for (GenTree* op : ops)
-            {
-                if (op == emitOp1)
-                {
-                    tgtPrefUse = BuildUse(op);
-                    srcCount++;
-                }
-                else if (op == emitOp2 || op == emitOp3)
-                {
-                    srcCount += BuildDelayFreeUses(op, emitOp1);
-                }
-            }
-
-            srcCount += BuildDelayFreeUses(intrin.op3, emitOp1);
+            prefUseNode = emitOp1;
```

Comment:
Don't we have the problem here for the `resultOpNum == 3` case? `prefUseNode` will end up being `intrinEmb.op3`, hence the last ref position we build can end up being a non-delay freed version of an earlier operand.

Reply:
AIUI, the code in `hwintrinsiccodegenarm64.cpp` will choose the correct instruction based on which input register matches the output register.

For the `resultOpNum == 3` case it knows `op3` is the last use so ideally that should be the RMW operand. Then lsra ensures that is the one that does _not_ get delay freed.

Possible there is something I'm missing here.


If that logic is correct, then I think the code can be simplified to:

```c++
            unsigned resultOpNum =
                embOp2Node->GetResultOpNumForRmwIntrinsic(user, intrinEmb.op1, intrinEmb.op2, intrinEmb.op3);

            if (resultOpNum == 0)
            {
                   prefUseNode = embOp2Node->Op(1);
            }
            else
            {
                  assert(resultOpNum >= 1 && resultOpNum <= 3);
                 prefUseNode = embOp2Node->Op(resultOpNum);
            }
```

Reply:
What I'm not understanding is _when_ do we have the correctness requirement that the destination register is not the same as an operand register? Does this case have the correctness requirement?

I am less interesting in the optimizations happening here until we get the constraints encoded correctly in a way that I know register allocation will handle. Currently I am not convinced, but I also do not have a good understanding of what cases have the register requirements we have been discussing.

Reply:
The requirements here would be that 1) `op1` and the destination are _ideally_ same register for RMW use 2) The other inputs can't use the same operand as the destination (because it might follow a `MOVPRFX`)

The codegen then outputs extra code to handle the case when `op1` and the destination don't match. Plus it asserts if other inputs share the same register as the destination.

The optimisation is that any of the inputs can be chosen as the RMW node (which is what the `GetResultOpNumForRmwIntrinsic()` is picking). Codegeneration spots this has happened by looking at which input matches the destination.

Reply:
Does anything prevent `prefUseNode` to be `op3` and `op2` to be the same interval as `op3`? I will create a delay-freed `op2` which has no effect. It will create non delay-freed `op3`. Then the destination register is allowed to be the same as `op2` and `op3`. If I understand correctly, this will hit the same problem as in this PR.

Reply:
```
 var vr3 = vr2.F1;
 var vr4 = vr2.F1;
 var vr5 = vr2.F1;
 vr2.F1 = Sve.FusedMultiplyAdd(vr3, vr4, vr5);
```

```
Generating: N033 (  1,  1) [000060] -----------                   t60 =    LCL_VAR   simd16 V08 cse0         u:1 d16 REG d16 <l:$2c1, c:$2c2>
Generating: N035 (  1,  1) [000062] -----------                   t62 =    LCL_VAR   simd16 V08 cse0         u:1 d16 REG d16 <l:$2c0, c:$300>
Generating: N037 (  1,  1) [000063] -----------                   t63 =    LCL_VAR   simd16 V08 cse0         u:1 d16 (last use) REG d16 <l:$2c0, c:$300>
Generating: N039 (???,???) [000081] -----------                   t81 =    HWINTRINSIC mask   double CreateTrueMaskAll REG p0
IN000b:             ptrue   p0.d
Generating: N041 (???,???) [000082] -c---------                   t82 =    CNS_VEC   simd16<0x00000000, 0x00000000, 0x00000000, 0x00000000> REG NA
                                                                        /--*  t60    simd16
                                                                        +--*  t62    simd16
                                                                        +--*  t63    simd16
Generating: N043 (  8,  7) [000039] -A-XG+-----                   t39 = *  HWINTRINSIC simd16 double FusedMultiplyAdd REG NA <l:$2c5, c:$2c6>
                                                                        /--*  t81    mask
                                                                        +--*  t39    simd16
                                                                        +--*  t82    simd16
Generating: N045 (???,???) [000083] -A-XG------                   t83 = *  HWINTRINSIC simd16 double ConditionalSelect REG d17
							V08 in reg d16 is becoming dead  [000063]
							Live regs: 0001000000000001 {x0 d16} - {d16} => 0000000000000001 {x0}
							Live vars after [000063]: {V03 V08} -{V08} => {V03}
IN000c:             movprfx z17.d, p0/z, z16.d
IN000d:             fmla    z17.d, p0/m, z16.d, z16.d
```

This works because the codegeneration for `FusedMultiplyAdd` can handle cases where the registers are the same.

That wasn't the case for the non-FMA intrinsic in the original fuzzlyn test.

Reply:
Which means we've got an inconsistency between different types of intrinsics in the codegen. Feels like the non-FMA intrinsic needs to be handling these cases as well? And the lsra should just be trying to give the most optimal register layout? ie the lsra doesn't have a correctness requirement?

Reply:
Aha, that's because the conditionalselect node which wraps it has it's own destination register:
```
[000083] -A-XG------                   t83 = *  HWINTRINSIC simd16 double ConditionalSelect REG d17
``

And the `FusedMultiplyAdd` is contained within the `ConditionalSelect`, so codegen uses the target register of the  `ConditionalSelect`.

Reply:
Curiously, this is exactly the two issues in #106867. Which this PR now fixes:

With the first:
```
Generating: N017 (  1,  1) [000028] -----------                   t28 =    LCL_VAR   simd16 V07 cse0         u:1 d17 REG d17 <l:$280, c:$2c0>
Generating: N019 (  3,  2) [000005] -----+-----                    t5 =    CNS_VEC   simd16<0x3f800000, 0x00000000, 0x00000000, 0x00000000> REG d18 $181
IN0009:             ldr     q18, (LARGELDC)[@RWD16]
Generating: N021 (  1,  1) [000030] -----------                   t30 =    LCL_VAR   simd16 V07 cse0         u:1 d17 (last use) REG d17 <l:$280, c:$2c0>
Generating: N023 (???,???) [000033] -----------                   t33 =    HWINTRINSIC mask   float CreateTrueMaskAll REG p1
IN000a:             ptrue   p1.s
Generating: N025 (???,???) [000034] -c---------                   t34 =    CNS_VEC   simd16<0x00000000, 0x00000000, 0x00000000, 0x00000000> REG NA
                                                                        /--*  t28    simd16 
                                                                        +--*  t5     simd16 
                                                                        +--*  t30    simd16 
Generating: N027 ( 12, 19) [000012] -A--G+-----                   t12 = *  HWINTRINSIC simd16 float FusedMultiplySubtractNegated REG NA <l:$300, c:$301>
                                                                        /--*  t33    mask   
                                                                        +--*  t12    simd16 
                                                                        +--*  t34    simd16 
Generating: N029 (???,???) [000035] -A--G------                   t35 = *  HWINTRINSIC simd16 float ConditionalSelect REG d17
							V07 in reg d17 is becoming dead  [000030]
							Live regs: 0002000000000001 {x0 d17} - {d17} => 0000000000000001 {x0}
							Live vars after [000030]: {V00 V07} -{V07} => {V00}
IN000b:             movprfx z17.s, p1/z, z17.s
IN000c:             fnmsb   z17.s, p1/m, z18.s, z17.s
```

is now:
```
Generating: N017 (  1,  1) [000028] -----------                   t28 =    LCL_VAR   simd16 V07 cse0         u:1 d17 REG d17 <l:$280, c:$2c0>
Generating: N019 (  3,  2) [000005] -----+-----                    t5 =    CNS_VEC   simd16<0x3f800000, 0x00000000, 0x00000000, 0x00000000> REG d18 $181
IN0009:             ldr     q18, (LARGELDC)[@RWD16]
Generating: N021 (  1,  1) [000030] -----------                   t30 =    LCL_VAR   simd16 V07 cse0         u:1 d17 (last use) REG d17 <l:$280, c:$2c0>
Generating: N023 (???,???) [000033] -----------                   t33 =    HWINTRINSIC mask   float CreateTrueMaskAll REG p1
IN000a:             ptrue   p1.s
Generating: N025 (???,???) [000034] -c---------                   t34 =    CNS_VEC   simd16<0x00000000, 0x00000000, 0x00000000, 0x00000000> REG NA
                                                                        /--*  t28    simd16 
                                                                        +--*  t5     simd16 
                                                                        +--*  t30    simd16 
Generating: N027 ( 12, 19) [000012] -A--G+-----                   t12 = *  HWINTRINSIC simd16 float FusedMultiplySubtractNegated REG NA <l:$300, c:$301>
                                                                        /--*  t33    mask   
                                                                        +--*  t12    simd16 
                                                                        +--*  t34    simd16 
Generating: N029 (???,???) [000035] -A--G------                   t35 = *  HWINTRINSIC simd16 float ConditionalSelect REG d19
							V07 in reg d17 is becoming dead  [000030]
							Live regs: 0002000000000001 {x0 d17} - {d17} => 0000000000000001 {x0}
							Live vars after [000030]: {V00 V07} -{V07} => {V00}
IN000b:             movprfx z19.s, p1/z, z17.s
IN000c:             fnmls   z19.s, p1/m, z18.s, z17.s
```

The second one: 
```
Generating: N017 (  1,  1) [000003] -----+-----                    t3 =    LCL_VAR   simd16<System.Numerics.Vector`1> V00 loc0         u:2 d16 REG d16 $100
Generating: N019 (  1,  1) [000004] -----+-----                    t4 =    LCL_VAR   simd16<System.Numerics.Vector`1> V00 loc0         u:2 d16 REG d16 $100
Generating: N021 (  1,  1) [000005] -----+-----                    t5 =    LCL_VAR   simd16<System.Numerics.Vector`1> V00 loc0         u:2 d16 (last use) REG d16 $100
Generating: N023 (???,???) [000025] -----------                   t25 =    HWINTRINSIC mask   short CreateTrueMaskAll REG p1
IN0006:             ptrue   p1.h
Generating: N025 (???,???) [000026] -c---------                   t26 =    CNS_VEC   simd16<0x00000000, 0x00000000, 0x00000000, 0x00000000> REG NA
                                                                        /--*  t3     simd16 
                                                                        +--*  t4     simd16 
                                                                        +--*  t5     simd16 
Generating: N027 (  4,  4) [000006] -c---+-----                    t6 = *  HWINTRINSIC simd16 short MultiplyAdd REG NA $200
                                                                        /--*  t25    mask   
                                                                        +--*  t6     simd16 
                                                                        +--*  t26    simd16 
Generating: N029 (???,???) [000027] -----------                   t27 = *  HWINTRINSIC simd16 short ConditionalSelect REG d16
							V00 in reg d16 is becoming dead  [000005]
							Live regs: 0001000000000000 {d16} - {d16} => 0000000000000000 {}
Debug: Closing V00 debug range.
							Live vars after [000005]: {V00} -{V00} => {}
IN0007:             movprfx z16.h, p1/z, z16.h
IN0008:             mad     z16.h, p1/m, z16.h, z16.h
```

is now:
```
Generating: N017 (  1,  1) [000003] -----+-----                    t3 =    LCL_VAR   simd16<System.Numerics.Vector`1> V00 loc0         u:2 d16 REG d16 $100
Generating: N019 (  1,  1) [000004] -----+-----                    t4 =    LCL_VAR   simd16<System.Numerics.Vector`1> V00 loc0         u:2 d16 REG d16 $100
Generating: N021 (  1,  1) [000005] -----+-----                    t5 =    LCL_VAR   simd16<System.Numerics.Vector`1> V00 loc0         u:2 d16 (last use) REG d16 $100
Generating: N023 (???,???) [000025] -----------                   t25 =    HWINTRINSIC mask   short CreateTrueMaskAll REG p1
IN0006:             ptrue   p1.h
Generating: N025 (???,???) [000026] -c---------                   t26 =    CNS_VEC   simd16<0x00000000, 0x00000000, 0x00000000, 0x00000000> REG NA
                                                                        /--*  t3     simd16 
                                                                        +--*  t4     simd16 
                                                                        +--*  t5     simd16 
Generating: N027 (  4,  4) [000006] -c---+-----                    t6 = *  HWINTRINSIC simd16 short MultiplyAdd REG NA $200
                                                                        /--*  t25    mask   
                                                                        +--*  t6     simd16 
                                                                        +--*  t26    simd16 
Generating: N029 (???,???) [000027] -----------                   t27 = *  HWINTRINSIC simd16 short ConditionalSelect REG d17
							V00 in reg d16 is becoming dead  [000005]
							Live regs: 0001000000000000 {d16} - {d16} => 0000000000000000 {}
Debug: Closing V00 debug range.
							Live vars after [000005]: {V00} -{V00} => {}
IN0007:             movprfx z17.h, p1/z, z16.h
IN0008:             mla     z17.h, p1/m, z16.h, z16.h
```